tsl/src/nodes/decompress_chunk/decompress_chunk.c

```diff
@@ -335,57 +327,105 @@ cost_decompress_chunk(Path *path, Path *compressed_path)
-	dcpath->custom_path.path.startup_cost = sort_path.total_cost;
+	/*
+	 * In compressed batch sorted merge, for each distinct segmentby value we
+	 * have to keep the corresponding latest batch open. Estimate the number of
+	 * these batches to account for it in the cost model.
+	 */
+	List *segmentby_groupexprs = NIL;
+	for (int segmentby_attno = bms_next_member(compression_info->chunk_segmentby_attnos, -1);
+		 segmentby_attno > 0;
+		 segmentby_attno =
+			 bms_next_member(compression_info->chunk_segmentby_attnos, segmentby_attno))
+	{
+		Var *var = palloc(sizeof(Var));
+		*var = (Var){ .xpr.type = T_Var,
+					  .varno = compression_info->compressed_rel->relid,
+					  .varattno = segmentby_attno };
+		segmentby_groupexprs = lappend(segmentby_groupexprs, var);
+	}
+
+	const double simultaneously_open_baches =
+		estimate_num_groups(root, segmentby_groupexprs, dcpath->custom_path.path.rows, NULL, NULL);
 
 	/*
-	 * The cost model for the normal chunk decompression produces the following total
-	 * costs.
-	 *
-	 * Segments  Total costs
-	 * 10         711.84
-	 * 50        4060.91
-	 * 100       8588.32
-	 * 10000   119281.84
-	 *
-	 * The cost model of the regular decompression is roughly linear. Opening multiple batches in
-	 * parallel needs resources and merging a high amount of batches becomes inefficient at some
-	 * point. So, we use a quadratic cost model here to have higher costs than the normal
-	 * decompression when more than ~100 batches are used. We set
-	 * DECOMPRESS_CHUNK_HEAP_MERGE_CPU_TUPLE_COST to 0.8 to become most costly as soon as we have to
-	 * process more than 120 batches.
-	 *
-	 * Note: To behave similarly to the cost model of the regular decompression path, this cost
-	 * model does not consider the number of tuples.
+	 * Getting one uncompressed row involves replacing the top row in the heap,
+	 * which costs O(log(heap size)).
+	 * The constant multiplier is found empirically by benchmarking queries returning
+	 * 1 - 1e9 tuples with segmentby cardinality 1 to 1e4, and adjusting the
+	 * cost so that the fastest plan is used.
 	 */
-	dcpath->custom_path.path.total_cost =
-		sort_path.total_cost + pow(sort_path.rows, 2) * DECOMPRESS_CHUNK_HEAP_MERGE_CPU_TUPLE_COST;
+	const double uncompressed_row_cost = 1.5 * log(simultaneously_open_baches) * cpu_tuple_cost;
 
-	dcpath->custom_path.path.rows = sort_path.rows * DECOMPRESS_CHUNK_BATCH_SIZE;
+	/*
+	 * Keeping a lot of batches open might use a lot of memory. The batch sorted
+	 * merge can't offload anything to disk, so we just penalize it heavily if
+	 * we expect it to go over the work_mem.
+	 */
+	const double work_mem_bytes = work_mem * 1024;
+	const double needed_memory_bytes = simultaneously_open_baches * DECOMPRESS_CHUNK_BATCH_SIZE *
+									   dcpath->custom_path.path.pathtarget->width;
+	const double work_mem_penalty =
+		0.5 * disable_cost *
+		smoothstep(needed_memory_bytes, 0.75 * work_mem_bytes, 1.25 * work_mem_bytes);
+
```

Comment:
Since you have an assert for the other components of the cost model, I suggest also adding

```
Assert(work_mem_penalty > 0);
```

to be consistent.